operatorcourier/api.py

```diff
@@ -41,42 +41,29 @@ def build_and_verify(source_dir=None, yamls=None, ui_validate_io=False,
     :raises OpCourierBadBundle: When the resulting bundle fails validation
     """
 
-    if source_dir is not None and yamls is not None:
-        logger.error("Both source_dir and yamls cannot be defined.")
-        raise TypeError(
-            "Both source_dir and yamls cannot be specified on function call.")
+    manifests = get_manifests_info(source_dir, yamls)
+    validation_dict, bundle_dict \
+        = get_validation_dict_from_manifests(manifests, ui_validate_io, repository)
 
-    yaml_files = []
-
-    if source_dir is not None:
-        for filename in os.listdir(source_dir):
-            if filename.endswith(".yaml") or filename.endswith(".yml"):
-                with open(source_dir + "/" + filename) as f:
-                    yaml_files.append(f.read())
-    elif yamls is not None:
-        yaml_files = yamls
-
-    bundle = BuildCmd().build_bundle(yaml_files)
+    nested = len(manifests) > 1
 
-    valid, validation_results_dict = ValidateCmd(ui_validate_io).validate(bundle,
-                                                                          repository)
+    bundle = None
+    if not nested:
+        bundle = format_bundle(bundle_dict)
 
-    if validation_output is not None:
+    # writing warnings/errors as file to disk
+    if validation_output:
         with open(validation_output, 'w') as f:
-            f.write(json.dumps(validation_results_dict) + "\n")
+            f.write(json.dumps(validation_dict) + "\n")
 
-    if valid:
-        bundle = format_bundle(bundle)
-        if output is not None:
+    if output:
+        if nested:
+            logging.warning('The manifests are valid, but the bundle is not output to '
+                            'disk because the directory structure is nested.')
+        else:
             with open(output, 'w') as f:
                 yaml.dump(bundle, f, default_flow_style=False)
-        return bundle
-    else:
-        logger.error("Bundle failed validation.")
-        raise OpCourierBadBundle(
-            "Resulting bundle is invalid, input yaml is improperly defined.",
-            validation_info=validation_results_dict
-        )
+    return bundle
```

Comment:
We need to what we are returning here and why. IMO this change needs to either be a breaking change, or the API needs to be refactored such that there is a separate method for the API in the nested case. My personal preference would be to refactor this method to return something other than `bundle` and break the API, since it is cleaner for courier to determine if the folder is nested or not. But @MartinBasti @jeremylinlin @awgreene please suggest alternatives here.

Either way, I don't think this should be as straightforward as it is defined here. This is very messy and will get confusing quickly.